cli/flox/src/utils/search.rs

```diff
@@ -87,108 +87,148 @@ pub(crate) fn construct_search_params(
     Ok(params)
 }
 
-/// Deduplicate and disambiguate display items.
+/// An intermediate representation of a search result used for rendering
+#[derive(Debug, PartialEq, Clone)]
+pub struct DisplayItem {
+    /// The input that the package came from
+    input: String,
+    /// The attribute path of the package, excluding subtree and system
+    rel_path: Vec<String>,
+    /// The package description
+    description: Option<String>,
+    /// Whether to join the `input` and `package` fields with a separator when rendering
+    render_with_input: bool,
+}
+
+impl Display for DisplayItem {
+    /// Render a display item in the format that should be output by
+    /// `flox search`.
+    ///
+    /// It should be possible to copy and paste this as an argument to
+    /// `flox install`.
+    ///
+    /// If we change this function, we will likely need to update what the
+    /// deduplicate field controls in pkgdb.
+    /// Technically, pkgdb shouldn't have knowledge of this format,
+    /// but it's nicer to perform deduplication in SQL.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.render_with_input {
+            write!(f, "{}{SEARCH_INPUT_SEPARATOR}", self.input)?;
+        }
+        write!(f, "{}", self.rel_path.join("."))
+    }
+}
+
+/// Contains [DisplayItem]s that have been disambiguated.
 ///
-/// This gets complicated because we have to satisfy a few constraints:
-/// - The order of results from `pkgdb` is important (best matches come first),
-///   so that order must be preserved.
-/// - Versions shouldn't appear in the output, so multiple package versions from a single
-///   input should be deduplicated.
-/// - Packages that appear in more than one input need to be disambiguated by prepending
-///   the name of the input and a separator.
-fn dedup_and_disambiguate_display_items(mut display_items: Vec<DisplayItem>) -> Vec<DisplayItem> {
-    let mut package_to_inputs: HashMap<String, HashSet<String>> = HashMap::new();
-    for d in display_items.iter() {
-        // Build a collection of packages and which inputs they are seen in so we can tell
-        // which packages need to be disambiguated when rendering search results.
-        package_to_inputs
-            .entry(d.package.clone())
-            .and_modify(|inputs| {
-                inputs.insert(d.input.clone());
+/// This should be used for printing search results when the format output by
+/// [DisplaySearchResults] is not desired.
+pub struct DisplayItems(pub Vec<DisplayItem>);
```

Comment:
I'm not a huge fan of the proliferation of different kinds of search result structs, but wasn't sure of a better way to do this

Reply:
I think it's fairly idiomatic... 

Not sure if we _need_ the `impl Display for DisplayItem` since we don't know whether we need to render with input for any given display item
and we then also print the description in a whole other display implementation after all)

Reply:
What's the whole other display implementation we print it in?

Reply:
It's not used in `impl Display for DisplayItem` its also not used in `impl Display for DisplayItems` only two layers down in `impl Display for SearchResultItems`.
that last impl makes use of both `impl Display for DisplayItem` and individual fields (`description`) which is kinda weird.

Reply:
That is a little weird. I think having printing in a single place for package descriptors is good so that search and install don't diverge, especially since changes to what we display will require changes in pkgdb. Having it in one place seems like it improves our odds of not introducing a regression

Reply:
yeah, maybe move the display logic for the display item into `impl Display for SearchResultItems` just to have all that in one place at least

Reply:
But we use it in `DidYouMean` when we don't have `SearchResultItems`:
```
        let display_items: DisplayItems = self.search_results.results.clone().into();
        for result in display_items.iter() {
            writeln!(f, "  $ flox install {result}",)?;
        }
```